runtime-modules/project-token/src/lib.rs

```diff
@@ -800,6 +801,239 @@ decl_module! {
+            if amount.is_zero() {
+                return Ok(()); // noop
+            }
+
+            let sender = ensure_signed(origin.clone())?;
+
+            T::MemberOriginValidator::ensure_member_controller_account_origin(
+                origin,
+                member_id
+            )?;
+
+            let token_data = Self::ensure_token_exists(token_id)?;
+            let curve = token_data.bonding_curve.ok_or(Error::<T>::NotInAmmState)?;
+            let user_acc_data = Self::ensure_account_data_exists(token_id, &member_id)?;
+
+            ensure!(user_acc_data.amount >= amount, Error::<T>::UnsufficientTokenAmount);
+
+            let amm_reserve_account = Self::module_bonding_curve_reserve_account(token_id);
+
+            let price = curve.eval::<T>(amount, token_data.total_supply, BondOperation::Unbond)?;
+
+            Self::ensure_can_transfer_joy(&amm_reserve_account, price)?;
+
+            // slippage tolerance check
+            if let Some((slippage_tolerance, desired_price)) = slippage_tolerance {
+                ensure!(desired_price.saturating_sub(price) < slippage_tolerance.mul_floor(desired_price), Error::<T>::SlippageToleranceExceeded);
+            }
+
+            // timestamp deadline check
+            if let Some(deadline) = deadline {
+                ensure!(<timestamp::Pallet<T>>::get() < deadline, Error::<T>::DeadlineExpired);
+            }
+
+            // == MUTATION SAFE ==
+
+            AccountInfoByTokenAndMember::<T>::mutate(token_id, member_id, |account_data| {
+                account_data.amount = account_data.amount.saturating_sub(amount)
+            });
+
+            TokenInfoById::<T>::mutate(token_id, |token_data| {
+                token_data.total_supply = token_data.total_supply.saturating_sub(amount);
+            });
+
+            Self::transfer_joy(&amm_reserve_account, &sender, price)?;
+
+            Self::deposit_event(RawEvent::TokenUnbonded(token_id, member_id, amount, price));
+
+            Ok(())
+        }
+
+        /// Deactivate the amm functionality
+        /// Preconditions
+        /// - (origin, member_id) must be a valid authentication pair
+        /// - token_id must be a valid
+        /// - token must be in `BondigCurve` state
+        ///
+        /// Postconditions
+        /// - Bonding Curve set to None
+        /// - state set to idle
+        /// - event deposited
+        #[weight = 10_000_000] // TODO: Adjust weight
+        fn deactivate_amm(origin, token_id: T::TokenId, member_id: T::MemberId) -> DispatchResult {
+            T::MemberOriginValidator::ensure_member_controller_account_origin(
+                origin,
+                member_id
+            )?;
+
+            let token_data = Self::ensure_token_exists(token_id)?;
+
+            ensure!(token_data.creator_member_id == member_id, Error::<T>::UserNotAuthorized);
+
+            ensure!(OfferingStateOf::<T>::ensure_bonding_curve_of::<T>(&token_data).is_ok(), Error::<T>::NotInAmmState);
```

Comment:
`ensure_bonding_curve_of` already returns `Error::<T>::NotInAmmState` on failure:
```suggestion
            OfferingStateOf::<T>::ensure_bonding_curve_of::<T>(&token_data)?;
```